ui/volumes: Store metrics timespan in URL

alexis-ld opened the pull request:
**Component**: ui

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

We want to store the Volume metrics timespan in the URL as a param query to allow users to share the URL while specifying a timespan.

**Summary**:

On timespan selection we write the param query to the URL following the specs described in the issue : 
* Last 7days: `&from=now-7d`
* Last 24hours: `&from=now-24h`
* Last 1 hour: `&from=now-1h`


On metrics card component load, we check if a `from` param query is defined. If it is set we dispatch an Update Volume Stats redux action followed by a data fetch.

To handle URL query strings parsing and generation I used the `query-string` package ( https://www.npmjs.com/package/query-string ). I found this package to be reliable and well-maintained (over 12M weekly downloads), however I can still switch to a custom handmade URL query param system if needed.

**Acceptance criteria**: 
* Changing a timespan on the Volume metrics updates the URL with query params
* On page load if a valid query param is set, the timespan is changed accordingly
* If other query params are set they are preserved 



---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2754 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Commit:
ui/volumes: Store metrics timespan in URL

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented:
Hello @alexis-ld!

> To handle URL query strings parsing and generation I used the `query-string` package

I think we used [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) somewhere else in our code, wouldn't it be sufficient?

alexis-ld commented:
@gdemonet Indeed. I missed that. Updating ... :)

alexis-ld commented:
@gdemonet Updated and squashed :)

ChengYanJin reviewed (changes requested):
Hello @alexis-ld,

I found there are two issues related to this PR:

1. When we switch from one volume to another, we would like to keep the same timespan selected not only in the dropdown but also in the URL.
2. When I set the URL with `?from=now-1h` I don't see the graphs with data anymore. Could you please investigate what is the problem here? 

Thank you.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

Commit:
ui/volumes: Fix race condition in volume metrics fetch

Commit:
ui/volumes:
Keep metric timespan in url even when volumes change

alexis-ld commented:
@ChengYanJin 

Thanks for the feedback. Updated 

1. URL Timespan is now preserved when changing the selected volume.

2. I investigated the data disappearing when loading the page on  -1h timespan.

This occurred because the VolumePage component was launching a metrics fetch on mount with basic params. Then quickly after that the VolumeMetricGraphCard was launching a fetch when seeing a valid URL query param. This created a race condition. On -1h the dataset being way smaller than the default 24h the requests resolved before the previously fired request resulting in random behavior.

The main reason behind this was the use of `takeEvery` on metrics related saga watchers. Since we want the metrics corresponding to the latest user action and not the ones that resolve first (race condition) I changed both metrics related saga watchers to `takeLatest` to avoid concurrency and it fixed everything nicely 😄  .

ChengYanJin reviewed (approved):
Thanks for investigating the issue and fixing it. 
Good job!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

alexis-ld commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexis-ld.
